### ISSUE TITLE
Ability to advance the Runtime or LocalSet until all tasks are Pending?

### ISSUE THREAD

**Issue by EkardNT** [tokio-rs/tokio#2443]
I would like to use tokio in a game I am making to implement coroutines that can execute across multiple game ticks. The way I would like this to work is that each tick of the main game loop, I would tell tokio's runtime to process all woken tasks until all futures have returned Pending, aka there is no more work to do at the present time. Then I would proceed with the rest of the game's update logic for that loop iteration. On the next loop iteration, the game would advance tokio's runtime again in the same manner. This would keep happening over and over until the program exits.

This might only make sense to do with a LocalSet, given that the full Runtime could potentially have multiple threads. There are also possibilities for more advanced control knobs around the basic all_pending stop condition, such as advancing until `all_pending || num_tasks_processed > some_configurable_threshold`, or having a timeout, etc. These could be useful in the context of a game where you want to tightly control the runtime of a single frame.

One possibility I've considered for achieving this using tokio's existing feature set is to call `LocalSet::run_until` with a future that simply awaits a yield_now call before completing.

```rust
// Totally untested code
fn every_game_tick() {
    runtime.block_on(local_set.run_until(async {
        yield_now().await;
    }));
}
```
The idea is that the yield_now() call would put the future at the back of the queue, allowing any other pending tasks to execute first. The problem with this approach is that it only does one pass through tokio's list of tasks, which may not be enough. For example, if one of the tasks spawns another one, I'm not sure whether this solution would cause that new task to be polled during the current game tick.

**Comment by Darksonn** [comment]
This is probably not the best way to go around that. Using Tokio in a game can be a decent way to handle network connections, but it should probably go on a different thread than the main game loop.

**Comment by carllerche** [comment]
There is currently no way to tick the runtime until all tasks are pending.

Assuming you are using I/O, one problem is, how do you wake up your game loop when an I/O event has been received?

**Comment by thombles** [comment]
I would find this feature very useful for simplified testing. I've spent considerable time trying alternative strategies to know when things are "done" but they've all proven unsatisfactory one way or another.

With this you could:
1. Send into a channel then await quiescence. When that returns you know the value has been received _and processed_, without going to pains to expose that or being forced to use a fixed time delay.
2. Advance mock time and know that the task that was waiting for it has completed, as have any cascading wakeups or tasks spawned off that one.

Conventionally, for testing without real delays you would mock out all this async stuff and interrogate each system synchronously but I feel like there's a real opportunity here for testing the async system as-is. Being able to wait for an idle runtime is a blunt tool but I keep coming back to it as the easiest way to reach this particular nirvana.

If this is not an appropriate thing for the real tokio Runtime (for good reasons I'm doubtless ignorant of), is there a way we could build such a thing ourselves out of the tools tokio gives us?

**Comment by dggmez** [comment]
@thombles @EkardNT Switch to [smol](https://github.com/smol-rs/smol). It includes two functions `try_tick` and `tick` to accomplish what you guys want:

https://docs.rs/smol/1.2.4/smol/struct.LocalExecutor.html#method.try_tick

https://docs.rs/smol/1.2.4/smol/struct.LocalExecutor.html#method.tick

**Comment by Darksonn** [comment]
There's no need to switch executor just to use [`LocalExecutor`][2] as a replacement for `LocalSet`. You can do that through the [`async-executor`][1] crate, and its `tick` method can be given to `Runtime::block_on` on a Tokio runtime and it will work similarly a Tokio `LocalSet`.

[1]: https://docs.rs/async-executor/
[2]: https://docs.rs/async-executor/1.3.0/async_executor/struct.LocalExecutor.html

**Comment by hawkw** [comment]
> @thombles @EkardNT Switch to [smol](https://github.com/smol-rs/smol). It includes two functions `try_tick` and `tick` to accomplish what you guys want:
> 
> https://docs.rs/smol/1.2.4/smol/struct.LocalExecutor.html#method.try_tick
> 
> https://docs.rs/smol/1.2.4/smol/struct.LocalExecutor.html#method.tick

I don't think that `smol::LocalExecutor`'s `tick` and `try_tick` methods actually have the desired behavior. The documentation for `LocalExecutor::tick` states (emphasis mine):

> Run **a single task**.
>
> Running a task means simply polling its future once.

This suggests that `LocalExecutor::tick` will poll _the next task_ in the local executor's run queue, if one is scheduled. This is different from polling _all_ tasks in the run queue until they are all pending, which is the behavior described above.

It's also worth noting Tokio's `LocalSet` implements `Future`, and its `Future::poll` implementation polls tasks spawned on the `LocalSet`: 
https://github.com/tokio-rs/tokio/blob/e309da0beeb48a0c2ecd0f15eb600f871ec98e19/tokio/src/task/local.rs#L459-L477

To advance the `LocalSet` until all tasks on it are pending, you could poll the `LocalSet` future.

Note that there is a limit on the number of tasks that the `LocalSet` will poll per tick: https://github.com/tokio-rs/tokio/blob/e309da0beeb48a0c2ecd0f15eb600f871ec98e19/tokio/src/task/local.rs#L217

Therefore, if you want to ensure that _all_ tasks on the `LocalSet` are polled to `Pending`, you may need to track the number of tasks spawned on the `LocalSet` and poll it multiple times if that number is greater than the number of tasks polled per tick.

**Comment by piegamesde** [comment]
Tokio not being able to integrate with external event loops is a real show stopper :(

One thing that would already help me a lot would be a `tokio::runtime::Runtime::block_on_async` so that I can easily wrap futures in a Tokio context while keeping it single-threaded.

**Comment by notgull** [comment]
Seconding this. Having a method like [`Executor::run()`](https://docs.rs/async-executor/latest/async_executor/struct.Executor.html#method.run) on the `Runtime` or `Handle` type would be a boon for integration.

**Comment by alshdavid** [comment]
Also need this functionality to be able to embed Tokio into an external event loop.

**Comment by coder137** [comment]
By using `smol` [AsyncTask](https://github.com/smol-rs/async-task) have written a small module here: https://github.com/coder137/ticked-async-executor

The only limitation with `TickedAsyncExecutor` is that it does not have access to the tokio runtime internally, and cannot use tokio libraries which depend on that runtime (for example the Timer functionality etc)

This has been useful for me when I want to run/test an async task one tick at a time and synchronized with the main thread.

Feel free to review and suggest how the tokio runtime can be added to this

**Comment by alshdavid** [comment]
Likewise I made something similar [here](https://github.com/alshdavid-forks/napi-rs/tree/alsh/async_local/crates/napi/src/async_local).

I am using an adapted single threaded futures executor from the `futures` crate. I have an off-thread waker that ensures the main thread never blocks on waiting for futures to wake.

**Comment by Noah-Kennedy** [comment]
I am currently working on a patch set that provide this functionality.

**Comment by sylveon** [comment]
Did this patch set ever happen? APIs I'm using require my thread to run a Windows message loop (though I can control that loop). It would be perfect if I could get some kind of callback when an event is received, that way I could set a Win32 event (that the loop would wait on using `MsgWaitForMultipleObjectsEx`). Then the loop code could `tick()` when that event is set.
